### PR TITLE
feat: support specifying a turbo version in turbo.json

### PR DESCRIPTION
**Pull request by chris-olszewski**
Add support and update schema for specifying `turboVersion` in `turbo.json`. We also verify that the listed version is an exact version and that turbo isn't listed as a dev dependency since the presence of `turboVersion` implies the use of global turbo.

I also updated our semver library to 3 in order to use their exact semver function. I think this is the version we want as one of the primary breaking changes was to modify how `^` is handled to be closer to `npm` and `Cargo`: https://github.com/Masterminds/semver/blob/master/CHANGELOG.md#changed-1.

Added some tests to make sure that the verification logic errors as expected.

### CODE REVIEW

File: cli/internal/fs/turbo_json_test.go [hunk 1]
```diff
@@ -245,6 +246,40 @@ func Test_TaskOutputsSort(t *testing.T) {
...
+	packageJSONPath := testDir.UntypedJoin("package.json")
+	rootPackageJSON, pkgJSONReadErr := ReadPackageJSON(packageJSONPath)
+
+	if pkgJSONReadErr != nil {
+		t.Fatalf("invalid parse: %#v", pkgJSONReadErr)
+	}
+
+	_, turboJSONReadErr := ReadTurboConfig(testDir, rootPackageJSON)
+
+	expectedErrorMsg := "Found versions for turbo in both turbo.json and root package.json. Please delete the package.json dependency"
```
Comment (mehulkar) [inline cli/internal/fs/turbo_json_test.go:278]: > Please delete the package.json dependency

Is it possible to make this more actionable like: "Please run `npm uninstall -D turbo`" (except `npm` is the detected package manager) and `-D` is based on whether it's a dependency or devDependency).

If package manager/copy-pastable command is too far of a stretch, maybe just:

> "Please remove "turbo" from `dependencies`"

or

> "Please remove "turbo" from `devDependencies`"

File: packages/turbo-types/src/types/config.ts [hunk 2]
```diff
@@ -3,6 +3,15 @@ export interface Schema {
   /** @default https://turbo.build/schema.json */
   $schema?: string;
 
+  /**
+   * A version of turbo that should be used with this project
+   *
+   * If turbo is invoked in a project with turboVersion specified, turbo will make
+   * sure to use the correct version. Only semver versions that include patch
+   * numbers are allowed.
```
Comment (mehulkar) [inline packages/turbo-types/src/types/config.ts:11]: - Are canaries allowed? 
- One of the things I really dislike about corepack is that I have to choose an exact version. Most of the time, I don't know what minor or patch versions are even available, and I can never remember how/where to find that info. As a maintainer of turbo, I'd want exact version also, but as a user, I would really just want `1.x` or at most `1.5.x`

File: cli/go.mod [hunk 3]
```diff
@@ -5,7 +5,7 @@ go 1.18
 require (
 	github.com/AlecAivazis/survey/v2 v2.3.5
 	github.com/DataDog/zstd v1.5.2
-	github.com/Masterminds/semver v1.5.0
```
Comment (mehulkar) [inline cli/go.mod:8]: Super lazy of me not to google this, but what's the mechanism to update the installed package when you want to do this? Is it just `go get github.com/Masterminds/semver`?

File: packages/turbo-types/src/types/config.ts [hunk 4]
```diff
@@ -3,6 +3,15 @@ export interface Schema {
...
+   */
+  turboVersion?: string;
```
Comment (mehulkar) [inline packages/turbo-types/src/types/config.ts:13]: Name bikeshed: could it just be "version"? Or is that too ambiguous

File: cli/internal/context/context.go [hunk 5]
```diff
@@ -15,7 +15,7 @@ import (
 	"github.com/vercel/turbo/cli/internal/turbopath"
 	"github.com/vercel/turbo/cli/internal/util"
 
-	"github.com/Masterminds/semver"
+	"github.com/Masterminds/semver/v3"
```
Comment (mehulkar) [inline cli/internal/context/context.go:18]: Newb Go question... why does this import change? We just updated the dependency version right, shouldn't it just take what's in `go.mod`?

File: cli/internal/fs/turbo_json.go [hunk 6]
```diff
@@ -179,6 +183,10 @@ func ReadTurboConfig(rootPath turbopath.AbsoluteSystemPath, rootPackageJSON *Pac
 			rootPackageJSON.LegacyTurboConfig = nil
 		}
 
+		if hasPackageJSONTurboVersion && turboJSON.TurboVersion != "" {
+			return nil, fmt.Errorf("Found versions for turbo in both turbo.json and root package.json. Please delete the package.json dependency")
+		}
```
Comment (mehulkar) [inline cli/internal/fs/turbo_json.go:188]: What happens if there's a `turbo` key in `package.json`? test case for that might also be good. Eg. if package.json is:

```json
{
	"turbo": {
		"turboVersion": "1.7.0"
	}
}
```

File: cli/internal/fs/turbo_json.go [hunk 7]
```diff
@@ -355,5 +363,14 @@ func (c *TurboJSON) UnmarshalJSON(data []byte) error {
 	c.Pipeline = raw.Pipeline
 	c.RemoteCacheOptions = raw.RemoteCacheOptions
 
+	if raw.TurboVersion != "" {
+		// possibly strict here
+		version, err := semver.StrictNewVersion(raw.TurboVersion)
```
Comment (mehulkar) [inline cli/internal/fs/turbo_json.go:368]: might not make a difference perf-wise but is this more than a regex match of `/\d+\.\d+\.\d+/`? (see canaries question in another comment also)

File: cli/go.mod (hunk 3 above)
Comment (chris-olszewski) [reply cli/go.mod:8]: Go gets funky with major version bumps so you need to do `go get github.com/Masterminds/semver/v3` and it's viewed as a new package. If it was a non-major version change you can just do `go get github.com/Masterminds/semver@1.6.0`

File: cli/internal/context/context.go (hunk 5 above)
Comment (chris-olszewski) [reply cli/internal/context/context.go:18]: Go is weird about major versions: https://go.dev/doc/modules/major-version

File: packages/turbo-types/src/types/config.ts (hunk 2 above)
Comment (chris-olszewski) [reply packages/turbo-types/src/types/config.ts:11]: > Are canaries allowed?

[Yup](https://github.com/Masterminds/semver/blob/v3.1.1/version.go#L195)! Prereleases and build metadata is allowed.

> I would really just want 1.x or at most 1.5.x

It's annoying, but I really think we need to be strict here. Say `turboVersion` is `1.5.x` and Dev 1 gets a failure due to a bug in turbo `1.5.0`, Dev 2 tries to help so they clone the repo and run the same command, but we've pushed a patch that solves it so they don't see it and causes more confusion. The update notifier will help with this scenario, but it still means that a team can have version drift.

I think the better solution would be to have a codemod available that can resolve a version constraint or tag to a concrete version. e.g. `npx @turbo/codemod upgrade-turbo 1.5`

**Review by tknickman — COMMENTED** [review]
I think we need to make it more clear that `turboVersion` only impact `turbo` when _globally_ installed.

File: packages/turbo-types/src/types/config.ts (hunk 4 above)
Comment (chris-olszewski) [reply packages/turbo-types/src/types/config.ts:13]: I think being explicit is cheap here. To me `version` could seem like a version of either the `turbo.json` schema itself or the monorepo itself. Not a strong opinion though.

**Comment by chris-olszewski** [comment]
I'm still Go brained, this should really be happening in Rust

**Pull request closed without merging**